Implement Optional<T&>, use it to reduce copying

BenWiederhake opened the pull request:
This continues the work of #11142: Make fewer pointless copies of "heavy" objects. In particular, `Optional<T&>` did not exist, making the `HashMap::get` interface unnecessarily complicated and copy-heavy. This PR introduces a new `HashMap::get_ref` which returns a `Option<V&>`, to avoid copying the returned value. This is particularly useful if we just want to modify a single value of an entry.

In theory, every use of `HashMap::get` can be replaced by `HashMap::get_ref().copy()`, which would then allow us to get rid of `::get` and the `Trait::PeekType` mechanism. However, such a change is too far-reaching for now, so this PR only introduces the tools necessary to avoid copying Vectors and HashMaps.

I found these instances by introducing a new `bool Traits<T>::is_heavy()` class method, and returning true for Vectors, HashMaps, ByteBuffers. I decided against including it in this PR, but you can view it here: https://github.com/BenWiederhake/serenity/tree/historic/detect-hashmap-copy
Feel free to read, use, modify, and/or redistribute the code.

creator1creeper1 commented:
Nice changes. However, I think `Optional<T&>` should have the semantics of `T&`, so in particular assignment should be deep and not shallow. 
```
    ALWAYS_INLINE Optional& operator=(T& value)
    {
        m_ptr = &value;
        return *this;
    }
```
should be
```
    ALWAYS_INLINE Optional& operator=(T& value)
    {
        *m_ptr = value;
        return *this;
    }
```
IMO. However, this is a difficult question. Could we not simply delete the assignment operators, since their semantics are rather confusing?

BenWiederhake commented:
Changes since last summary:
- Rebased
- Fix up implicit merge conflict in Kernel::CommandLine
- Remove `operator=(T const&)`, since it is not absolutely clear what should happen. Note that `operator=(Optional<T&> const&)` continues to be implicit, and will copy the pointer (and hence *not* call `T::operator=(T const&)`).

creator1creeper1 commented:
> * Remove `operator=(T const&)`, since it is not absolutely clear what should happen. Note that `operator=(Optional<T&> const&)` continues to be implicit, and will copy the pointer (and hence _not_ call `T::operator=(T const&)`).

I still don't agree with these semantics. Consider the following:
```c++
int a = 3;
int b = 5;
int& r = a;
r = b;
```
vs.
```c++
int a = 3;
int b = 5;
Optional<int&> r = a;
r = b;
```
IMO, these should behave the same way. However, in the first case, there is deep assigment (because references are simply aliases for the referred-to object), but in the second case, there is shallow assignment (because `operator=` is `default`ed, and so the pointer is simply copied).
[Godbolt Link](https://godbolt.org/z/We6v6e3vG)

BenWiederhake commented:
Oof, I thought that deleting `operator=(T const&)` would get rid of this topic, but I forgot about the implicit constructor.

Hmm, how to resolve this?
- I think both shallow-assign and deep-assign are weird in some situations, hence why I wanted to remove it entirely.
- I think `operator=(Optional<T> const&)` really should do a shallow-assign, otherwise we get into all kinds of weird trouble (does deep-assigning to an empty Optional do a VERIFY, or set the Optional (which would be a deep-assign)? I would call both of these surprising.)
- Writing `operator=(T const&) = delete; operator=(T&) = delete;` will cause the compiler to reject this attempt even before the implicit constructor is attempted. Maybe that's the way to resolve this issue? What do you think?

Changes since last push:
- Add the `operator=(T const&) = delete; operator=(T&) = delete;` and adjust the tests for it (turns out, I accidentally used `operator=(T const&)` in one of the tests)
- Rebased

creator1creeper1 commented:
No, I don't think you realize the severity of the issue. `operator=` between `Optional<T&>` has the exact same issue:
```c++
int a = 3;
int b = 5;
int& a_r = a;
int& b_r = b;
a_r = b_r;
```
=> `a: 5, b: 5, a_r: 5, b_r: 5`
vs.
```c++
int a = 3;
int b = 5;
Optional<int&> a_r = a;
Optional<int&> b_r = b;
a_r = b_r;
```
=> `a: 3, b: 5, a_r: 5, b_r: 5`
[Godbolt Link](https://godbolt.org/z/fhvh456eG)

BenWiederhake commented:
Well, yes, that's what I wrote: Optional<T&> should have shallow-assign, obviously.

creator1creeper1 commented:
But that leads to the exampe I have shown, which I think is counter-intuitive. But you make a good point too: We can't always deep-assign, because when the assigned-to Optional is empty, we can only shallow-assign. Which is why I think there shouldn't be assignment at all.

creator1creeper1 commented:
As I already said: Your use-cases don't require assignment at all. I think it's best to explicitely `delete` *all* assignment operators on `Optional<T&>`.

BenWiederhake commented:
I don't think removing copy-assign is a good idea, but it seems to be the only way forward.

Changes since last push:
- `Optional& operator=(T const&) = delete;`
- Rebased

creator1creeper1 commented on AK/Optional.h:
```diff
@@ -185,6 +185,119 @@ class [[nodiscard]] Optional {
+    ALWAYS_INLINE void emplace(T& parameter)
+    {
+        m_ptr = parameter;
+    }
```
I can see that this compiles, but why? How can we assign a reference to a pointer?

creator1creeper1 commented on AK/Optional.h:
```diff
@@ -185,6 +185,119 @@ class [[nodiscard]] Optional {
+    [[nodiscard]] ALWAYS_INLINE T* value_ptr()
+    {
+        return m_ptr;
+    }
```
Reasonable addition to the usual Optional API.

creator1creeper1 commented on Kernel/CommandLine.cpp:
```diff
@@ -86,9 +86,9 @@ UNMAP_AFTER_INIT CommandLine::CommandLine(const String& cmdline_from_bootloader)
+StringView CommandLine::lookup(StringView key, StringView fallback) const
 {
-    return m_params.get(key);
+    return m_params.get_ref(key).value_or(fallback);
```
Are we really avoiding a copy here if we're returning by value anyway?

creator1creeper1 commented on Kernel/Devices/DeviceManagement.cpp:
```diff
@@ -45,10 +45,7 @@ DeviceManagement& DeviceManagement::the()
-        if (it == map.end())
-            return nullptr;
-        return it->value;
+        return map.get_ref(encoded_device(major, minor)).value_or(nullptr);
```
`.value_ptr()`?

creator1creeper1 commented on AK/Optional.h:
```diff
@@ -185,6 +185,119 @@ class [[nodiscard]] Optional {
+        m_ptr = other.m_ptr;
+        other.m_ptr = nullptr;
+        return *this;
+    }
```
I know I'm very insisting about this, but please `delete` this as well.
Consider the following:
```c++
int a = 3;
int b = 5;
int& a_r = a;
int& b_r = b;
a_r = std::move(b_r);
```
=> `a: 5, b: 5, a_r: 5, b_r: 5`
```c++
int a = 3;
int b = 5;
Optional<int&> a_r = a;
Optional<int&> b_r = b;
```
=> `a: 3, b: 5, a_r: 5, b_r: {}`
Now, the state of `b_r` doesn't matter because it has just been moved from. What matters is that `a_r` in the first example is deeply assigned to, changing the state of `a`, while in the second example, `a_r` is only shallowly assigned to, not changing `a`. Once again, I think that *if* we were to allow this operation, it should be deep-if-possible. 
[Godbolt Link](https://godbolt.org/z/czh1h1j9a)

BertalanD replied:
Throwing this here for the sake of completeness: the semantically equivalent `std::optional<std::reference_wrapper<T>>` also behaves like this PR in that it does shallow copies:
```cpp
int a = 3;
int b = 5;
std::optional<std::reference_wrapper<int>> a_r = a;
std::optional<std::reference_wrapper<int>> b_r = b;
a_r = std::move(b_r);
```
=> `a: 3, b: 5, a_r: 5, b_r: 5`

[Godbolt link](https://godbolt.org/z/8vfoEEPKT)

creator1creeper1 commented:
I fundamentally agree with the changes this PR makes. Please keep pushing for it to get merged!

creator1creeper1 commented on AK/Optional.h:
```diff
@@ -185,6 +185,119 @@ class [[nodiscard]] Optional {
+
+    ALWAYS_INLINE Optional() = default;
+
+    ALWAYS_INLINE Optional(Optional const& other) = delete;
```
You didn't need to delete this one. Reference creation is shallow as well, so there can be no inconsistency here. The inconsistensies can only arise when we try to change the state of the Optional after it has been created.

creator1creeper1 commented:
I fundamentally agree with the changes this PR makes. Just keep all the constructors and remove all the `operator=`s. Please keep pushing for this to get merged!

BenWiederhake commented:
I lost the motivation to work on this topic. Feel free to use any code I wrote.

creator1creeper1 commented:
It's not my goal to annoy you even more, but after discussing this a little more on the discord and thinking about it even more, there might also be lifetime issues with this. `T&` extends lifetimes of temporaries while `Optional<T&>` doesn't.
Consider the code at the [Godbolt Link](https://godbolt.org/z/MhEc7Tojq).

BenWiederhake commented:
Oh. Yup, that is a serious problem. I wish good luck to whoever takes on this problem.
